packages/perspective-jupyterlab/src/ts/renderer.ts

```diff
@@ -84,13 +85,39 @@ export class PerspectiveDocumentWidget extends DocumentWidget<PerspectiveWidget>
 
             if (this._psp.viewer.table === undefined) {
                 // construct new table
-                this._psp.viewer.load(WORKER.table(data));
+                const table = await WORKER.table(data);
+
+                // load data
+                this._psp.viewer.load(table);
+
+                // create a flat view
+                const view = await table.view();
+                view.on_update(() => {
+                    if (this._type === "csv") {
+                        view.to_csv().then((result: string) => {
+                            this.context.model.fromString(result);
+                            this.context.save();
+                        });
+                    } else if (this._type === "arrow") {
+                        view.to_arrow().then((result: ArrayBuffer) => {
+                            const resultAsB64 = btoa(new Uint8Array(result).reduce((acc, i) => (acc += String.fromCharCode.apply(null, [i])), ""));
+                            this.context.model.fromString(resultAsB64);
```

Comment:
What does this do - this creates an arrow and not a base64 encoded string in a text file?

Reply:
Yep the key is in a different code block where we tell jupyter to open arrow files as binary which means they're exposed in the model as b64 encoded strings https://github.com/finos/perspective/blob/master/packages/perspective-jupyterlab/src/ts/renderer.ts#L181